File: app/src/main/java/com/kakacat/minitool/common/ui/RecycleViewScrollListener.java
package com.kakacat.minitool.common.ui;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.recyclerview.widget.StaggeredGridLayoutManager;

import com.kakacat.minitool.common.util.UiUtil;

public class RecycleViewScrollListener extends RecyclerView.OnScrollListener{

    private int lastItemPosition;
    private int layoutManagerType;
    private int[] lastPositions;
    private OnSwipeUpRefresh onSwipeUpRefresh;

    public RecycleViewScrollListener(OnSwipeUpRefresh onSwipeUpRefresh) {
        lastItemPosition = 0;
        layoutManagerType = -1;
        this.onSwipeUpRefresh = onSwipeUpRefresh;
    }

    @Override
    public void onScrollStateChanged(@NonNull RecyclerView recyclerView, int newState) {
        RecyclerView.LayoutManager layoutManager = recyclerView.getLayoutManager();
        assert layoutManager != null;
        if(layoutManagerType == -1){
            if(layoutManager instanceof LinearLayoutManager){
                layoutManagerType = LayoutManagerType.LinearLayoutManager;
            }else if(layoutManager instanceof GridLayoutManager){
                layoutManagerType = LayoutManagerType.GridLayoutManager;
            }else if(layoutManager instanceof StaggeredGridLayoutManager){
                layoutManagerType = LayoutManagerType.StaggeredGridLayoutManager;
            }else{
                throw new RuntimeException("UnSupport LayoutManager");
            }
        }
        switch (layoutManagerType){
            case LayoutManagerType.LinearLayoutManager:{
                lastItemPosition = ((LinearLayoutManager) layoutManager).findLastVisibleItemPosition();
                break;
            }
            case LayoutManagerType.GridLayoutManager:{
                lastItemPosition = ((GridLayoutManager)layoutManager).findLastVisibleItemPosition();
                break;
            }
            case LayoutManagerType.StaggeredGridLayoutManager:{
                StaggeredGridLayoutManager manager = (StaggeredGridLayoutManager)layoutManager;
                if(lastPositions == null){
                    lastPositions = new int[manager.getSpanCount()];
                }
                manager.findLastVisibleItemPositions(lastPositions);
                for (int lastPosition : lastPositions) {
                    lastItemPosition = Math.max(lastPosition,lastItemPosition);
                }
                break;
            }
            default:
                break;
        }

        if(onSwipeUpRefresh != null){
            if(newState == RecyclerView.SCROLL_STATE_IDLE && layoutManager.getChildCount() > 0 && lastItemPosition >= layoutManager.getItemCount() -1){
                showLoading(recyclerView);
                onSwipeUpRefresh.loadMore();
            }
        }
    }

    private void showLoading(@NonNull RecyclerView recyclerView){
        UiUtil.showLoading(recyclerView.getContext(),recyclerView);
    }

    public interface OnSwipeUpRefresh{
        void loadMore();
    }

    private interface LayoutManagerType{
        int LinearLayoutManager = 0;
        int GridLayoutManager = 1;
        int StaggeredGridLayoutManager = 2;
    }
}

File: app/src/main/java/com/kakacat/minitool/common/ui/RecycleViewScrollListener.kt
package com.kakacat.minitool.common.ui

import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.StaggeredGridLayoutManager
import com.kakacat.minitool.common.util.UiUtil.showLoading

class RecycleViewScrollListener(private val onSwipeUpRefresh: OnSwipeUpRefresh) : RecyclerView.OnScrollListener() {

    private var lastItemPosition = 0
    private var layoutManagerType: Int = -1
    private var lastPositions: IntArray? = null

    override fun onScrollStateChanged(recyclerView: RecyclerView, newState: Int) {
        val layoutManager = recyclerView.layoutManager!!
        if (layoutManagerType == -1) {
            when (layoutManager) {
                is LinearLayoutManager -> {
                    LayoutManagerType.LinearLayoutManager
                }
                is GridLayoutManager -> {
                    LayoutManagerType.GridLayoutManager
                }
                is StaggeredGridLayoutManager -> {
                    LayoutManagerType.StaggeredGridLayoutManager
                }
                else -> {
                    throw RuntimeException("UnSupport LayoutManager")
                }
            }
        }
        when (layoutManagerType) {
            LayoutManagerType.LinearLayoutManager -> {
                lastItemPosition = (layoutManager as LinearLayoutManager).findLastVisibleItemPosition()
            }
            LayoutManagerType.GridLayoutManager -> {
                lastItemPosition = (layoutManager as GridLayoutManager).findLastVisibleItemPosition()
            }
            LayoutManagerType.StaggeredGridLayoutManager -> {
                val manager = layoutManager as StaggeredGridLayoutManager
                if (lastPositions == null) {
                    lastPositions = IntArray(manager.spanCount)
                }
                manager.findLastVisibleItemPositions(lastPositions)
                for (lastPosition in lastPositions!!) {
                    lastItemPosition = lastPosition.coerceAtLeast(lastItemPosition)
                }
            }
            else -> {
            }
        }
        if (newState == RecyclerView.SCROLL_STATE_IDLE && layoutManager.childCount > 0 && lastItemPosition >= layoutManager.itemCount - 1) {
            showLoading(recyclerView)
            onSwipeUpRefresh.loadMore()
        }
    }

    private fun showLoading(recyclerView: RecyclerView) {
        showLoading(recyclerView.context, recyclerView)
    }

    interface OnSwipeUpRefresh {
        fun loadMore()
    }

    private interface LayoutManagerType {
        companion object {
            const val LinearLayoutManager = 0
            const val GridLayoutManager = 1
            const val StaggeredGridLayoutManager = 2
        }
    }
}